### PR TITLE
Object reference updates (#825)

### PR DESCRIPTION
**Pull request by tgiphil**
See #825

### CODE REVIEW

**Comment by tgiphil** [comment]
This commit changes object references to point to object's first field.

File: Source/Mosa.Compiler.Framework/Delegate.cs [hunk 1]
```diff
@@ -11,14 +11,14 @@ namespace Mosa.Compiler.Framework
 	/// <summary>
 	/// Patches delegates
 	/// </summary>
-	public static class DelegatePatcher
+	public static class Delegate
```
Comment (charsleysa) [inline Source/Mosa.Compiler.Framework/Delegate.cs:14]: this name change seems likely to create confusion with System.Delegate, would recommend keeping the `DelegatePatcher` name

File: Source/Mosa.Compiler.Framework/MethodScheduler.cs [hunk 2]
```diff
@@ -43,21 +36,21 @@ public sealed class MethodScheduler
...
 		/// </value>
-		public int TotalQueuedMethods => _totalQueued;
+		public int TotalQueuedMethods => totalQueued;
 
 		#endregion Properties
 
 		public MethodScheduler(Compiler compiler)
 		{
-			_channel = Channel.CreateUnbounded<CompilerResult>(new UnboundedChannelOptions { SingleWriter = false, SingleReader = false });
+			//_channel = Channel.CreateUnbounded<MosaMethod>(new UnboundedChannelOptions { SingleWriter = false, SingleReader = false });
```
Comment (charsleysa) [inline Source/Mosa.Compiler.Framework/MethodScheduler.cs:53]: remove leftover code

File: Source/Mosa.Compiler.Framework/MosaTypeLayout.cs [hunk 3]
```diff
@@ -31,27 +31,27 @@ public class MosaTypeLayout
 		/// <summary>
 		/// Holds the method table slot index
 		/// </summary>
-		private readonly Dictionary<MosaMethod, int> methodSlots = new Dictionary<MosaMethod, int>(new MosaMethodFullNameComparer());
+		private readonly Dictionary<MosaMethod, int> methodSlots = new Dictionary<MosaMethod, int>();
```
Comment (charsleysa) [inline Source/Mosa.Compiler.Framework/MosaTypeLayout.cs:34]: shouldn't `int` be changed to `uint` to match other changes?

File: Source/Mosa.Compiler.Framework/MosaTypeLayout.cs [hunk 4]
```diff
@@ -914,7 +911,8 @@ private int FindOverrideSlot(List<MosaMethod> methodTable, MosaMethod method)
 
 			foreach (var baseMethod in methodTable)
 			{
-				if (baseMethod.Name.Equals(method.Name) && baseMethod.Equals(method))
+				//if (baseMethod.Name.Equals(method.Name) && baseMethod.Equals(method))
```
Comment (charsleysa) [inline Source/Mosa.Compiler.Framework/MosaTypeLayout.cs:914]: remove leftover code

File: Source/Mosa.Compiler.Framework/OpcodeEncoder.cs [hunk 5]
```diff
@@ -402,6 +402,7 @@ public void Append64BitImmediate(Operand operand)
 			}
 			else
 			{
+				Debug.Assert(operand.Offset == 0);  // REMOVE ME
```
Comment (charsleysa) [inline Source/Mosa.Compiler.Framework/OpcodeEncoder.cs:405]: remove?

File: Source/Mosa.Compiler.Framework/Stages/CILDecodingStage.cs [hunk 6]
```diff
@@ -287,6 +286,42 @@ private void InsertExceptionStartInstructions()
 			}
 		}
 
+		protected bool IsSourceAndTargetWithinSameTryOrException(InstructionNode node)
```
Comment (charsleysa) [inline Source/Mosa.Compiler.Framework/Stages/CILDecodingStage.cs:289]: resolve warning:
`'CILDecodingStage.IsSourceAndTargetWithinSameTryOrException(InstructionNode)': new protected member declared in sealed type`

File: Source/Mosa.Compiler.Framework/Stages/CILProtectedRegionStage.cs [hunk 7]
```diff
@@ -69,9 +69,9 @@ private void UpdateBlockProtectInstructions()
 
 						// Find enclosing try or finally handler
 						var exceptionHandler = FindImmediateExceptionHandler(headerBlock.Label);
-						bool isTry = exceptionHandler.IsLabelWithinTry(headerBlock.Label);
+						bool InTry = exceptionHandler.IsLabelWithinTry(headerBlock.Label);
```
Comment (charsleysa) [inline Source/Mosa.Compiler.Framework/Stages/CILProtectedRegionStage.cs:72]: variable should be camelCase

File: Source/Mosa.Demo.TestWorld.x86/Mosa.Demo.TestWorld.x86.csproj [hunk 8]
```diff
@@ -1,4 +1,36 @@
 ﻿<Project Sdk="Microsoft.NET.Sdk">
+	<PropertyGroup>
```
Comment (charsleysa) [inline Source/Mosa.Demo.TestWorld.x86/Mosa.Demo.TestWorld.x86.csproj:2]: looks like a merge issue, all these "new" lines should be removed

File: Source/Mosa.Korlib/System/BitConverter.cs [hunk 9]
```diff
@@ -19,81 +19,74 @@ public static byte[] GetBytes(bool value)
...
+		// Converts a short into an array of bytes with length two.
+		public unsafe static byte[] GetBytes(short value)
 		{
-			byte[] bytes = new byte[sizeof(short)];
-			Unsafe.As<byte, short>(ref bytes[0]) = value;
+			//byte[] bytes = new byte[sizeof(short)];
+			//Unsafe.As<byte, short>(ref bytes[0]) = value;
+			//return bytes;
+			byte[] bytes = new byte[2];
+			fixed (byte* b = bytes)
```
Comment (charsleysa) [inline Source/Mosa.Korlib/System/BitConverter.cs:32]: prefer the use of Unsafe helper class instead

File: Source/Mosa.Korlib/System/BitConverter.cs [hunk 10]
```diff
@@ -19,81 +19,74 @@ public static byte[] GetBytes(bool value)
...
 
 		// Converts an int into an array of bytes with length
 		// four.
-		public static byte[] GetBytes(int value)
+		public unsafe static byte[] GetBytes(int value)
 		{
-			byte[] bytes = new byte[sizeof(int)];
-			Unsafe.As<byte, int>(ref bytes[0]) = value;
+			byte[] bytes = new byte[4];
+			fixed (byte* b = bytes)
```
Comment (charsleysa) [inline Source/Mosa.Korlib/System/BitConverter.cs:42]: prefer the use of Unsafe helper class instead

File: Source/Mosa.Korlib/System/BitConverter.cs [hunk 11]
```diff
@@ -19,81 +19,74 @@ public static byte[] GetBytes(bool value)
...
 
 		// Converts a long into an array of bytes with length
 		// eight.
-		public static byte[] GetBytes(long value)
+		public unsafe static byte[] GetBytes(long value)
 		{
-			byte[] bytes = new byte[sizeof(long)];
-			Unsafe.As<byte, long>(ref bytes[0]) = value;
+			byte[] bytes = new byte[8];
+			fixed (byte* b = bytes)
```
Comment (charsleysa) [inline Source/Mosa.Korlib/System/BitConverter.cs:52]: prefer the use of Unsafe helper class instead

File: Source/Mosa.Korlib/System/BitConverter.cs [hunk 12]
```diff
@@ -19,81 +19,74 @@ public static byte[] GetBytes(bool value)
...
 
 		// Converts a double into an array of bytes with length
 		// eight.
-		public static byte[] GetBytes(double value)
+		public unsafe static byte[] GetBytes(double value)
 		{
-			byte[] bytes = new byte[sizeof(double)];
-			Unsafe.As<byte, double>(ref bytes[0]) = value;
-			return bytes;
+			return GetBytes(*(long*)&value);
```
Comment (charsleysa) [inline Source/Mosa.Korlib/System/BitConverter.cs:89]: prefer the use of Unsafe helper class instead

File: Source/Mosa.Korlib/System/BitConverter.cs [hunk 13]
```diff
@@ -19,81 +19,74 @@ public static byte[] GetBytes(bool value)
...
 
 		// Converts a float into an array of bytes with length
 		// four.
-		public static byte[] GetBytes(float value)
+		public unsafe static byte[] GetBytes(float value)
 		{
-			byte[] bytes = new byte[sizeof(float)];
-			Unsafe.As<byte, float>(ref bytes[0]) = value;
-			return bytes;
+			return GetBytes(*(int*)&value);
```
Comment (charsleysa) [inline Source/Mosa.Korlib/System/BitConverter.cs:82]: prefer the use of Unsafe helper class instead

File: Source/Mosa.Korlib/System/BitConverter.cs [hunk 14]
```diff
@@ -19,81 +19,74 @@ public static byte[] GetBytes(bool value)
...
 		}
 
 		// Converts an unsigned long into an array of bytes with
 		// length eight.
 		public static byte[] GetBytes(ulong value)
 		{
-			byte[] bytes = new byte[sizeof(ulong)];
-			Unsafe.As<byte, ulong>(ref bytes[0]) = value;
-			return bytes;
+			return GetBytes((long)value);
```
Comment (charsleysa) [inline Source/Mosa.Korlib/System/BitConverter.cs:75]: prefer the use of Unsafe helper class instead

File: Source/Mosa.Korlib/System/BitConverter.cs [hunk 15]
```diff
@@ -19,81 +19,74 @@ public static byte[] GetBytes(bool value)
...
 		}
 
 		// Converts an uint into an array of bytes with
 		// length four.
 		public static byte[] GetBytes(uint value)
 		{
-			byte[] bytes = new byte[sizeof(uint)];
-			Unsafe.As<byte, uint>(ref bytes[0]) = value;
-			return bytes;
+			return GetBytes((int)value);
```
Comment (charsleysa) [inline Source/Mosa.Korlib/System/BitConverter.cs:68]: prefer the use of Unsafe helper class instead

File: Source/Mosa.Korlib/System/BitConverter.cs [hunk 16]
```diff
@@ -19,81 +19,74 @@ public static byte[] GetBytes(bool value)
...
 		}
 
 		// Converts an ushort into an array of bytes with
 		// length two.
 		public static byte[] GetBytes(ushort value)
 		{
-			byte[] bytes = new byte[sizeof(ushort)];
-			Unsafe.As<byte, ushort>(ref bytes[0]) = value;
-			return bytes;
+			return GetBytes((short)value);
```
Comment (charsleysa) [inline Source/Mosa.Korlib/System/BitConverter.cs:61]: prefer the use of Unsafe helper class instead

File: Source/Mosa.Korlib/System/BitConverter.cs [hunk 17]
```diff
@@ -19,81 +19,74 @@ public static byte[] GetBytes(bool value)
 		// Converts a char into an array of bytes with length two.
 		public static byte[] GetBytes(char value)
 		{
-			byte[] bytes = new byte[sizeof(char)];
-			Unsafe.As<byte, char>(ref bytes[0]) = value;
-			return bytes;
+			return GetBytes((short)value);
```
Comment (charsleysa) [inline Source/Mosa.Korlib/System/BitConverter.cs:22]: prefer the use of Unsafe helper class instead

File: Source/Mosa.Korlib/System/NotImplementedException.cs [hunk 18]
```diff
@@ -8,7 +8,7 @@ namespace System
 	[Serializable]
 	public class NotImplementedException : SystemException
 	{
-		private readonly string message;
+		//private readonly string message;
```
Comment (charsleysa) [inline Source/Mosa.Korlib/System/NotImplementedException.cs:11]: remove leftover code

File: Source/Mosa.UnitTests.x86/Mosa.UnitTests.x86.csproj [hunk 19]
```diff
@@ -1,4 +1,23 @@
 ﻿<Project Sdk="Microsoft.NET.Sdk">
+	<PropertyGroup>
```
Comment (charsleysa) [inline Source/Mosa.UnitTests.x86/Mosa.UnitTests.x86.csproj:2]: merge issue, all these "new" lines should be removed

**Review by charsleysa — CHANGES_REQUESTED** [review]

File: Source/Mosa.Compiler.Framework.API/Mosa.Compiler.Framework.API.csproj [hunk 20]
```diff
@@ -1,6 +1,6 @@
 ﻿<Project Sdk="Microsoft.NET.Sdk">
   <PropertyGroup>
-    <TargetFramework>net5.0</TargetFramework>
+    <TargetFramework>netstandard2.0</TargetFramework>
```
[the remark this answers is not in the document]
Comment (tgiphil) [reply Source/Mosa.Compiler.Framework.API/Mosa.Compiler.Framework.API.csproj:3]: Good catch. Must be an inaccurate merge at some point.

File: Source/Mosa.Compiler.Framework/Delegate.cs (hunk 1 above)
Comment (tgiphil) [reply Source/Mosa.Compiler.Framework/Delegate.cs:14]: Agreed!

File: Source/Mosa.Compiler.Common/StreamExtension.cs [hunk 21]
```diff
@@ -21,13 +21,12 @@ public static void Write(this Stream stream, byte[] buffer)
...
 		/// <param name="stream">The stream.</param>
 		/// <param name="size">The size.</param>
-		public static void WriteZeroBytes(this Stream stream, int size)
+		public static void WriteZeroBytes(this Stream stream, uint size)
 		{
 			for (int i = 0; i < size; i++)
 				stream.WriteByte(0);
 		}
 
-		/// <summary>
```
[the remark this answers is not in the document]
Comment (tgiphil) [reply Source/Mosa.Compiler.Common/StreamExtension.cs:30]: Check.

File: Source/Mosa.Compiler.Framework.API/Mosa.Compiler.Framework.API.csproj (hunk 20 above)
[the remark this answers is not in the document]
Comment (tgiphil) [reply Source/Mosa.Compiler.Framework.API/Mosa.Compiler.Framework.API.csproj:3]: Actually, leftover from an old code tree. Removed.

File: Source/Mosa.Compiler.Framework.xUnit/Mosa.Compiler.Framework.xUnit.csproj [hunk 22]
```diff
@@ -1,48 +1,49 @@
...
+	</PropertyGroup>
+	<PropertyGroup Condition=" '$(Configuration)|$(Platform)' == 'Release|AnyCPU' ">
+		<DebugSymbols>True</DebugSymbols>
+		<TreatWarningsAsErrors>true</TreatWarningsAsErrors>
+		<WarningLevel>2</WarningLevel>
+	</PropertyGroup>
+	<ItemGroup>
+		<PackageReference Include="Microsoft.NET.Test.Sdk" Version="16.8.3" />
+		<PackageReference Include="Microsoft.NET.Test.Sdk" Version="16.9.1" />
+		<PackageReference Include="Microsoft.NET.Test.Sdk" Version="16.11.0" />
```
[the remark this answers is not in the document]
Comment (tgiphil) [reply Source/Mosa.Compiler.Framework.xUnit/Mosa.Compiler.Framework.xUnit.csproj:30]: Check.

File: Source/Mosa.Compiler.Framework/Delegate.cs (hunk 1 above)
Comment (tgiphil) [reply Source/Mosa.Compiler.Framework/Delegate.cs:14]: Check.

File: Source/Mosa.Compiler.Framework/MethodScheduler.cs (hunk 2 above)
Comment (tgiphil) [reply Source/Mosa.Compiler.Framework/MethodScheduler.cs:53]: Check.

File: Source/Mosa.Korlib/System/BitConverter.cs (hunk 15 above)
Comment (tgiphil) [reply Source/Mosa.Korlib/System/BitConverter.cs:68]: Check.

File: Source/Mosa.Korlib/System/BitConverter.cs (hunk 16 above)
Comment (tgiphil) [reply Source/Mosa.Korlib/System/BitConverter.cs:61]: Check.

File: Source/Mosa.Korlib/System/BitConverter.cs (hunk 17 above)
Comment (tgiphil) [reply Source/Mosa.Korlib/System/BitConverter.cs:22]: Check.

File: Source/Mosa.Korlib/System/NotImplementedException.cs (hunk 18 above)
Comment (tgiphil) [reply Source/Mosa.Korlib/System/NotImplementedException.cs:11]: Check.

File: Source/Mosa.UnitTests.x86/Mosa.UnitTests.x86.csproj (hunk 19 above)
Comment (tgiphil) [reply Source/Mosa.UnitTests.x86/Mosa.UnitTests.x86.csproj]: Check.

**Review by charsleysa — APPROVED** [review]
LGTM

**Pull request merged**